lc0/src/neural/network_cudnn.cu

```diff
@@ -0,0 +1,836 @@
+}
+
+BNLayer::~BNLayer()
+{
+    cudaFree(means);
+    cudaFree(variances);
+}
+
+FCLayer::FCLayer(BaseLayer *ip, int C, int H, int W, bool relu, bool bias, bool tanh) :
+    BaseLayer(C, H, W, ip),
+    useRelu(relu),
+    useBias(bias),
+    useTanh(tanh)
+{
+    size_t weightSize = bpe * C * H * W * ip->getC() * ip->getH() * ip->getW();
+    size_t biasSize = bpe * C * H * W;
+    cudaMalloc(&weights, weightSize);
+    if (useBias)
+    {
+        cudaMalloc(&biases, biasSize);
+    }
+    else
+    {
+        biases = NULL;
+    }
+}
+
+void FCLayer::loadWeights(void *cpuWeight, void *cpuBias)
+{
+    size_t weightSize = bpe * C * H * W * input->getC() * input->getH() * input->getW();
+    cudaMemcpyAsync(weights, cpuWeight, weightSize, cudaMemcpyHostToDevice);
+    if (useBias)
+    {
+        size_t biasSize = bpe * C * H * W;
+        cudaMemcpyAsync(biases, cpuBias, biasSize, cudaMemcpyHostToDevice);
+    }
+}
+
+void FCLayer::eval(void *outputTensor, void *inputTensor, void *input2, void *scratch)
+{
+    float alpha = 1.0f, beta = 0.0f;
+    int numOutputs = C * H * W;
+    int numInputs = input->getC() * input->getH() * input->getW();
+
+    if (fp16)
+    {
+        // TODO: implement this!
+        assert(0);
+    }
+    else
+    {
+        cublasSgemm(g_cublas, CUBLAS_OP_T, CUBLAS_OP_N,
+            numOutputs, N, numInputs,
+            &alpha,
+            (float*)weights, numInputs,
+            (float*)inputTensor, numInputs,
+            &beta,
+            (float*)outputTensor, numOutputs);
+
+        if (useBias || useRelu || useTanh)
+        {
+            addVectors((float*)outputTensor, (float*)biases, (float*)outputTensor, numOutputs * N, numOutputs, numOutputs * N, useRelu, useTanh);
+        }
+    }
+
+}
+
+FCLayer::~FCLayer()
+{
+    cudaFree(weights);
+    cudaFree(biases);
+}
+
+
+namespace lczero {
+
+class CudnnNetwork;
+
+class CudnnNetworkComputation : public NetworkComputation {
+ public:
+  CudnnNetworkComputation(const CudnnNetwork* network) : m_network(network) {}
+
+  void AddInput(InputPlanes&& input) override 
+  {
+      raw_input.emplace_back(input);
+  }
+
+  void ComputeBlocking() override;
+
+  int GetBatchSize() const override { return raw_input.size(); }
+
+  float GetQVal(int sample) const override 
+  {
+      return out_val[sample];
+  }
+  float GetPVal(int sample, int move_id) const override 
+  {
+      return out_pol[sample][move_id];
+  }
+
+ private:
+     // input
+     std::vector<InputPlanes> raw_input;
+
+     static constexpr int V2_NUM_OUTPUT_POLICY = 1858;
+
+     // output (TODO: try using cudaHostAlloc to avoid the copy?)
+     float out_pol[MAX_BATCH_SIZE][V2_NUM_OUTPUT_POLICY];
+     float out_val[MAX_BATCH_SIZE];
+     float input_planes[MAX_BATCH_SIZE][kInputPlanes*8*8];
+
+     const CudnnNetwork* m_network;
+};
+
+// currently only one NN eval can happen a time (we can fix this if needed by allocating more memory)
+static std::mutex g_lock;
```

Comment:
That is also something that seems, that it should be inside CudnnNetwork class.

Reply:
right. done.